Added images and other changes (#13)
uptane/deployment-considerations: security_considerations.md

In order to prevent attackers who compromise the director repository from causing one type of ECU to install images intended for another type, an OEM and / or its suppliers SHOULD include certain information about images in the Targets metadata.

Suppose attackers have compromised the Director repository. If release counters are used and further mitigating steps are taken, they can not rollback software updates. Furthermore, without an additional key compromise, attackers cannot cause arbitrary software attacks on Primaries and full verification Secondaries. However, attackers can cause ECUs of one hardware type to install images intended for another hardware type. To use an analogy, this is similar to causing Linkedsys routers to install images intended for GetNear routers.

ECU identifiers (e.g., serial numbers) specified in the Targets metadata signed by the Director repository do not solve this problem, because: (1) they are used by the Director repository only to instruct which ECU should install which image, and (2) they are not specified in the Targets metadata signed on the Image repository, because it is impractical to list all ECU identifiers that pertain to an image.

> In order to avoid this problem, the Targets metadata about unencrypted images on the Image repository SHOULD always include the TargetsModule.Custom.hardwareIdentifier attribute. A hardware identifier allows an OEM and / or its suppliers to succinctly capture an entire class of ECUs without listing each of their ECU identifiers. Note that the OEM and / or its suppliers SHALL ensure that hardware identifiers are unique across different hardware types of ECUs, so that attackers who compromise the Director repository cannot cause ECUs of one type to install images intended for another type.

Reviewer: Not from your changes, but while I'm looking at it...

```suggestion
In order to avoid this problem, the Targets metadata about unencrypted images on the Image repository SHOULD always include the TargetsModule.Custom.hardwareIdentifier attribute. A hardware identifier allows an OEM and/or its suppliers to succinctly capture an entire class of ECUs without listing each of their ECU identifiers. Note that the OEM and/or its suppliers SHALL ensure that hardware identifiers are unique across different hardware types of ECUs, so that attackers who compromise the Director repository cannot cause ECUs of one type to install images intended for another type.
```

> Before an ECU installs a new image, it SHOULD always check the hardware type of the image so that attackers cannot cause any ECU to install randomly chosen images not intended for it.

Reviewer: "an ECU", right?

Author: Yes. I'll fix.

Author: Fixed in editing pass soon to be committed.

## Preventing rollback attacks in case of Director compromise

> On the[Exceptional Operations](https://github.com/uptane/deployment-considerations/blob/master/exceptional_operations.md) page, we discussed how an OEM and / or its suppliers SHOULD use release counters in order to prevent rollback attacks should the Director repository be compromised. To further limit these attacks in this scenario, the OEM and / or its suppliers SHOULD also use the following recommendations.

Reviewer: ```suggestion
On the [Exceptional Operations](https://github.com/uptane/deployment-considerations/blob/master/exceptional_operations.md) page, we discuss how an OEM and/or its suppliers SHOULD use release counters in order to prevent rollback attacks if the Director repository is compromised. To further limit these attacks in this scenario, the OEM and/or its suppliers SHOULD also use the following recommendations.
```

> First, they SHOULD diligently remove obsolete images from new versions of Targets metadata files uploaded to the Image repository. This can prevent attackers who compromise the Director repository from being able to choose these obsolete images for installation. This method has a downside in that it complicates the update process for vehicles that require an intermediate update step. For example, an ECU has previously installed image A, and C is the latest image it should install. However, the ECU should install image B before it installs C, and B has already been removed from the Targets metadata on the Image repository in order to prevent / limit rollback attacks. Thus, the OEM and / or its suppliers SHOULD balance between the requirement to remove obsolete images from the Targets metadata, and the need to support ECUs that require installation of these obsolete images before updating to the latest images.

Reviewer: ```suggestion
First, they SHOULD diligently remove obsolete images from new versions of Targets metadata files uploaded to the Image repository. This can prevent attackers who compromise the Director repository from being able to choose these obsolete images for installation. This method has a downside in that it complicates the update process for vehicles that require an intermediate update step. For example, an ECU has previously installed image A, and C is the latest image it should install. However, the ECU should install image B before it installs C, and B has already been removed from the Targets metadata on the Image repository in order to prevent or limit rollback attacks. Thus, the OEM and/or its suppliers SHOULD balance between the requirement to remove obsolete images from the Targets metadata, and the need to support ECUs that require installation of these obsolete images before updating to the latest images.
```

> Second, they SHOULD decrease the expiration timestamps in the Targets metadata uploaded to the Image repository. This is done so that these Targets metadata expire more quickly, thus preventing attackers who compromise the Director repository from being able to choose these obsolete images. This method also has a downside, since these Targets metadata now need to be updated more quickly. To prevent accidental freeze attacks, an ECU needs to be able to update both the time from the time server, and metadata from the Image repository.  In this unlikely event that the ECU is able to update metadata, but not the time, it can continue working with the previously installed image, but would be unable to update to the latest image. The Director repository can detect this unlikely event using the vehicle version manifest. In this case, the OEM MAY require the owner of the vehicle to diagnose the problem at the nearest dealership or authorized mechanic.

Reviewer: ```suggestion
Second, they SHOULD decrease the expiration timestamps in the Targets metadata uploaded to the Image repository. This is done so that these Targets metadata expire more quickly, thus preventing attackers who compromise the Director repository from being able to choose these obsolete images. This method also has a downside, since these Targets metadata now need to be updated more quickly. To prevent accidental freeze attacks, an ECU needs to be able to update both the time from the Time Server, and metadata from the Image repository.  In this unlikely event that the ECU is able to update metadata, but not the time, it can continue working with the previously installed image, but would be unable to update to the latest image. The Director repository can detect this unlikely event using the vehicle version manifest. In this case, the OEM MAY require the owner of the vehicle to diagnose the problem at the nearest dealership or authorized mechanic.
```

Furthermore, the director repository knows that B-1.0 and C-3.0 conflict with each other. The only thing the director repository can be certain of is that B has installed either B-1.0 or B-2.0, and C has installed either C-1.0 or C-2.0. Thus, the director repository SHOULD NOT send the third bundle to the vehicle, because B-1.0 from the first bundle may still be installed, which would conflict with C-3.0 from the third bundle.
Therefore, the director repository SHOULD NOT issue the third bundle, until it has received a vehicle version manifest from the vehicle that confirms that ECUs B and C have installed the second bundle, which is known to contain images that do not conflict with the third bundle.
In conclusion, the director repository SHOULD NOT issue a new bundle until it has received confirmation, using the vehicle version manifest, that no image known to have been installed conflicts with the new images in the new bundle.

> If the director repository is not able to update a vehicle for any reason, then it SHOULD raise the issue to the OEM.

Reviewer: Without empty lines between these lines, this all becomes one big block of text. I suggest organizing it into a few paragraphs.

Author: Yeah. I'll fix this. This is still a work in progress. I committed the suggestions you made so far.

Author: Ok. I actually broke this up into bullets. I think it works better.

Consider the following example, where attackers have compromised the Director repository. If certain mitigating steps have been taken, such as using release counters, they can not rollback software updates. Furthermore, without an additional key compromise, attackers cannot cause arbitrary software attacks on Primaries and full verification Secondaries. However, attackers can cause the ECUs of one hardware type to install images intended for another hardware type. To use an analogy, this is similar to causing [Linkedsys](https://www.linksys.com/us/) routers to install images intended for [NetGear](https://www.netgear.com/) routers.

Simply having ECU identifiers (e.g., serial numbers) specified in the Targets metadata signed by the Director repository do not solve this problem, because: (1) they are used by the Director repository only to instruct which ECU should install which image, and (2) they are not specified in the Targets metadata signed on the Image repository, because it is impractical to list all ECU identifiers that pertain to an image.

> In order to avoid this problem, the Targets metadata about unencrypted images on the Image repository SHOULD always include the TargetsModule.Custom.hardwareIdentifier attribute. A hardware identifier allows an OEM and/or its suppliers to succinctly capture an entire class of ECUs without listing each of their ECU identifiers. Note that the OEM and/or its suppliers SHALL ensure that hardware identifiers are unique across different hardware types of ECUs, so that attackers who compromise the Director repository cannot cause ECUs of one type to install images intended for another type.

Reviewer 2: ```suggestion
In order to avoid this problem, the custom Targets metadata about unencrypted images on the Image repository SHOULD always include hardware identifiers. A hardware identifier allows an OEM and/or its suppliers to succinctly capture an entire class of ECUs without listing each of their ECU identifiers. Note that the OEM and/or its suppliers SHALL ensure that hardware identifiers are unique across different hardware types of ECUs, so that attackers who compromise the Director repository cannot cause ECUs of one type to install images intended for another type.
```

## Preventing rollback attacks in case of Director compromise

> On the [Exceptional Operations](https://github.com/uptane/deployment-considerations/blob/master/exceptional_operations.md#rolling-back-software) page, we discuss how an OEM and/or its suppliers SHOULD use release counters in order to prevent rollback attacks in case of a Director repository compromise. To further limit the impact of such an attack scenario, the OEM and/or its suppliers SHOULD also use the following recommendations.

Reviewer 2: ```suggestion
In the [Exceptional Operations](https://github.com/uptane/deployment-considerations/blob/master/exceptional_operations.md#rolling-back-software) document, we discuss how an OEM and/or its suppliers SHOULD use release counters in order to prevent rollback attacks in case of a Director repository compromise. To further limit the impact of such an attack scenario, the OEM and/or its suppliers SHOULD also use the following recommendations.
```

## Broadcasting vs. unicasting metadata inside the vehicle
An implementation of Uptane MAY have a Primary unicast metadata to Secondaries. In this scenario, the Primary would send metadata separately to each Secondary. However, this has a downside. Network disruptions can cause ECUs to see different versions of metadata released by repositories at different times.

> In order to mitigate this problem, it is RECOMMENDED that a Primary use a broadcast network such as CAN, CAN FD, or Ethernet to transmit metadata to all of its Secondaries at the same time. Primaries MAY use the ECUModule.MetadataFiles message to do so. Note that this still does not guarantee that ECUs will not end up seeing different versions of metadata released by repositories at different times. This is because network traffic between Primaries and Secondaries may still get disrupted, especially if they are connected through intermediaries, such as gateways. Nevertheless, it should still be better than unicasting.

Reviewer: What is this `ECUModule.MetadataFiles` message?

Author: This term came directly out of the Deployment Google Doc.I'm afraid I can't answer the question and I'm not sure how to fix it. Can we merge the rest and then resolve this issue?

Reviewer 2: Let's just drop that sentence.

Author: Good solution.

> In order to mitigate this problem, it is RECOMMENDED that a Primary use a broadcast network such as CAN, CAN FD, or Ethernet to transmit metadata to all of its Secondaries at the same time. Primaries MAY use the ECUModule.MetadataFiles message to do so. Note that this still does not guarantee that ECUs will not end up seeing different versions of metadata released by repositories at different times. This is because network traffic between Primaries and Secondaries may still get disrupted, especially if they are connected through intermediaries, such as gateways. Nevertheless, it should still be better than unicasting.

Reviewer 2: ```suggestion
In order to mitigate this problem, it is RECOMMENDED that a Primary use a broadcast network such as CAN, CAN FD, or Ethernet to transmit metadata to all of its Secondaries at the same time. Note that this still does not guarantee that ECUs will not end up seeing different versions of metadata released by repositories at different times. This is because network traffic between Primaries and Secondaries may still get disrupted, especially if they are connected through intermediaries, such as gateways. Nevertheless, it should still be better than unicasting.
```

If an update is intended to be applied to a gateway itself, it should be updated either before or after (but not during) update operations to ECUs on the other side of the gateway. This can help to avoid the disruption described above.

## Managing dependencies and conflicts between ECUS

The *dependencies* for a given ECU is the set of other images that SHOULD also be installed on other ECUs in order for that image to work on a particular ECU. The *conflicts* for the same image and ECU is the set of other images that SHOULD not be installed on other ECUs if the ECU in question is to work correctly. *Dependency resolution* is the process of determining which versions of the latest images and their dependencies can be installed without conflicts.

### Checking dependencies and conflicts

There are three options for checking dependencies and conflicts:

> 1. **Only ECUs check dependencies and conflicts.** This information should be included in the Targets metadata on the Image repository, and should not add substantially to bandwidth costs. The upside is that, without offline keys, attackers cannot cause ECUs to fail to satisfy dependencies and prevent conflicts. The downside is that it can add to computational costs, because dependency resolution is generally an NP-hard problem. However, it is possible to control the computational costs if some constraints are imposed.

Reviewer 2: ```suggestion
1. **Only ECUs check dependencies and conflicts.** This information should be included in the Targets metadata on the Image repository, and should not add substantially to bandwidth costs. The upside is that, without offline keys, attackers cannot cause ECUs to fail to satisfy dependencies and prevent conflicts. The downside is that it can add to computational costs, because dependency resolution is generally [an NP-hard problem](https://research.swtch.com/version-sat). However, it is possible to control the computational costs if some constraints are imposed.
```

In order to avoid these problems, whenever possible OEMs and suppliers SHOULD use ASN.1 decoders that have been comprehensively tested via unit tests and fuzzing.

Furthermore, following best practices, we recommend that DER encoding is used instead of BER and CER, because DER provides a unique encoding of values.

## Balancing EEPROM performance and security

Many ECUs use EEPROM which, practically speaking, can be written to only for a limited number of times. This in turn can impose limits on how often these ECUs can be updated.

In order to analyze this problem, let us recap what new information should be downloaded in every software update cycle:

1. The Primary writes and sends the latest vehicle version manifest to the Director repository.
2. If a Time Server is used, all Secondaries write and send fresh tokens to the Primary. 

> 3. All ECUs download, verify, and write the latest downloaded time from the Time Server, or whatever device is used to provide the current accurate time.

Reviewer: Would "whatever source" be clearer than "whatever device"?

Author: Yes. I will fix that.
